### PR TITLE
Kavan/rate limiting

### PR DESCRIPTION
**Pull request by kvsikand**
Add rate limiting to published topics.
Simple, straighforward, seems to work

### CODE REVIEW

File: src/RosClientNode.hpp [hunk 1]
```diff
@@ -39,12 +41,20 @@ class RosClientNode : public QObject {
...
-    auto root_offset = encode<T>(fbb, msg, metadata);
-    fbb.Finish(flatbuffers::Offset<void>(root_offset));
-    Q_EMIT ros_message_encoded(QByteArray(
-        reinterpret_cast<const char*>(fbb.GetBufferPointer()), fbb.GetSize()));
+    auto& rate_info = rate_limits[to_topic];
+    auto curr_time = std::chrono::high_resolution_clock::now();
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(curr_time - rate_info.first);
+    double elapsed = duration.count() / 1000.0;
+
+    if (elapsed > rate_info.second) {
```
Comment (kvsikand) [inline src/RosClientNode.hpp:49]: Note: if we get topics more frequently than the rate limit, we simply drop them.
Is this correct? Seems better than sending them late but open to other options

Comment (joydeep-b) [reply src/RosClientNode.hpp:49]: Yes, this is the correct behavior: drop more frequent messages.

File: src/RosClientNode.hpp [hunk 2]
```diff
@@ -39,12 +41,20 @@ class RosClientNode : public QObject {
...
   void encode_ros_msg(
       const T& msg, const std::string& msg_type, const std::string& to_topic) {
-    flatbuffers::FlatBufferBuilder fbb;
-    auto metadata = encode_metadata(fbb, msg_type, to_topic);
-    auto root_offset = encode<T>(fbb, msg, metadata);
-    fbb.Finish(flatbuffers::Offset<void>(root_offset));
-    Q_EMIT ros_message_encoded(QByteArray(
-        reinterpret_cast<const char*>(fbb.GetBufferPointer()), fbb.GetSize()));
+    auto& rate_info = rate_limits[to_topic];
+    auto curr_time = std::chrono::high_resolution_clock::now();
```
Comment (joydeep-b) [inline src/RosClientNode.hpp:45]: Just to be specific, `const auto`. If you mark as const and accidentally modify later, the compiler will find the bug for you.

File: src/RosClientNode.hpp [hunk 3]
```diff
@@ -39,12 +41,20 @@ class RosClientNode : public QObject {
...
+    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(curr_time - rate_info.first);
```
Comment (joydeep-b) [inline src/RosClientNode.hpp:46]: `const`

File: src/RosClientNode.hpp [hunk 4]
```diff
@@ -39,12 +41,20 @@ class RosClientNode : public QObject {
...
+    double elapsed = duration.count() / 1000.0;
```
Comment (joydeep-b) [inline src/RosClientNode.hpp:47]: `const`

**Comment by kvsikand** [comment]
I did test the bandwidth usage from my local computer. The bag file I was using contained compressed image data as well as 2d lidar+odometry etc (no 3d lidar).. Even without rate limiting, it never got above ~6 Mbps, but with the rate limiting on, various parameters were able to limit to pretty much any value down to ~2.
So I'm confident that this does properly limit the bandwidth usage; it will just be a matter of adjusting the parameters correctly to get a desired bandwidth

File: src/config.example.hpp [hunk 5]
```diff
@@ -63,30 +63,30 @@ static void configure_msg_types(RosClientNode& cn) {
 
   // must send to status topic to list robot in webviz
   cn.register_local_msg_type<amrl_msgs::RobofleetStatus>(
-      "/status", webviz_constants::status_topic);
+      "/status", webviz_constants::status_topic, 1);
```
[the remark this answers is not in the document]
Comment (kvsikand) [reply src/config.example.hpp:66]: Yuup, but given the content of the status message I don't know that updates at a <1s rate make sense

**Pull request merged**